l1-contracts/contracts/bridge/ntv/NativeTokenVault.sol

```diff
@@ -0,0 +1,419 @@
+/// @dev Designed for use with a proxy for upgradability.
+abstract contract NativeTokenVault is INativeTokenVault, IAssetHandler, Ownable2StepUpgradeable, PausableUpgradeable {
+    using SafeERC20 for IERC20;
+
+    /// @dev Contract that stores the implementation address for token.
+    /// @dev For more details see https://docs.openzeppelin.com/contracts/3.x/api/proxy#UpgradeableBeacon.
+    IBeacon public bridgedTokenBeacon;
+
+    /// @dev The address of the WETH token.
+    address public immutable override WETH_TOKEN;
+
+    /// @dev L1 Shared Bridge smart contract that handles communication with its counterparts on L2s
+    IAssetRouterBase public immutable override ASSET_ROUTER;
+
+    /// @dev Maps token balances for each chain to prevent unauthorized spending across ZK chains.
+    /// This serves as a security measure until hyperbridging is implemented.
+    /// NOTE: this function may be removed in the future, don't rely on it!
+    mapping(uint256 chainId => mapping(address l1Token => uint256 balance)) public chainBalance;
+
+    /// @dev A mapping assetId => tokenAddress
+    mapping(bytes32 assetId => address tokenAddress) public tokenAddress;
+
+    /// @dev A mapping assetId => isTokenBridged
+    mapping(bytes32 assetId => bool bridged) public isTokenBridged; // kl todo should we have isTokenNativeInstead
+
+    /**
+     * @dev This empty reserved space is put in place to allow future versions to add new
+     * variables without shifting down storage in the inheritance chain.
+     * See https://docs.openzeppelin.com/contracts/4.x/upgradeable#storage_gaps
+     */
+    uint256[47] private __gap;
+
+    /// @notice Checks that the message sender is the bridgehub.
+    modifier onlyAssetRouter() {
+        if (msg.sender != address(ASSET_ROUTER)) {
+            revert Unauthorized(msg.sender);
+        }
+        _;
+    }
+
+    /// @dev Contract is expected to be used as proxy implementation.
+    /// @dev Disable the initialization to prevent Parity hack.
+    /// @param _wethToken Address of WETH on deployed chain
+    /// @param _assetRouter Address of assetRouter
+    constructor(address _wethToken, address _assetRouter) {
+        _disableInitializers();
+        ASSET_ROUTER = IAssetRouterBase(_assetRouter);
+        WETH_TOKEN = _wethToken;
+    }
+
+    /// @notice Registers tokens within the NTV.
+    /// @dev The goal was to allow bridging native tokens automatically, by registering them on the fly.
+    /// @notice Allows the bridge to register a token address for the vault.
+    /// @notice No access control is ok, since the bridging of tokens should be permissionless. This requires permissionless registration.
+    function registerToken(address _nativeToken) external {
+        if (_nativeToken == WETH_TOKEN) {
+            revert TokenNotSupported(WETH_TOKEN);
+        }
+        require(_nativeToken == ETH_TOKEN_ADDRESS || _nativeToken.code.length > 0, "NTV: empty token");
```

Comment:
btw, do we want to support "native" ETH token for L2 chains? Also, this sort of native token support may lead to situations when a chain with custom base token has two asset ids for its base token: from native vault here and the original asset id